File: packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/ManifestFactory.java
package expo.modules.updates.manifest;

import org.json.JSONException;
import org.json.JSONObject;

import expo.modules.updates.UpdatesConfiguration;

public class ManifestFactory {

  private static final String TAG = ManifestFactory.class.getSimpleName();

  public static Manifest getManifest(JSONObject manifestJson, ManifestResponse httpResponse, UpdatesConfiguration configuration) throws Exception {
    String expoProtocolVersion = httpResponse.header("expo-protocol-version", null);

    if (expoProtocolVersion == null) {
      return LegacyManifest.fromLegacyManifestJson(manifestJson, configuration);
    } else if (Integer.valueOf(expoProtocolVersion) == 0) {
      return NewManifest.fromManifestJson(manifestJson, httpResponse, configuration);
    } else {
      throw new Exception("Unsupported expo-protocol-version: " + expoProtocolVersion);
    }
  }

  public static Manifest getEmbeddedManifest(JSONObject manifestJson, UpdatesConfiguration configuration) throws JSONException {
    if (manifestJson.has("releaseId")) {
      return LegacyManifest.fromLegacyManifestJson(manifestJson, configuration);
    } else {
      return BareManifest.fromManifestJson(manifestJson, configuration);
    }
  }
}

File: packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/ManifestFactory.kt
package expo.modules.updates.manifest

import expo.modules.updates.UpdatesConfiguration
import expo.modules.manifests.core.BareManifest
import expo.modules.manifests.core.NewManifest
import org.json.JSONException
import org.json.JSONObject

/**
 * Utility methods for parsing a JSON manifest using the correct [UpdateManifest] implementation.
 */
object ManifestFactory {
  private val TAG = ManifestFactory::class.java.simpleName

  @Throws(Exception::class)
  fun getManifest(manifestJson: JSONObject, responseHeaderData: ResponseHeaderData, extensions: JSONObject?, configuration: UpdatesConfiguration): UpdateManifest {
    return when (val expoProtocolVersion = responseHeaderData.protocolVersion) {
      // TODO(wschurman): remove error in a few major releases after SDK 51 when it's unlikely classic updates
      // may erroneously be served
      null -> {
        throw Exception("Legacy manifests are no longer supported")
      }
      0, 1 -> {
        NewUpdateManifest.fromNewManifest(NewManifest(manifestJson), extensions, configuration)
      }
      else -> {
        throw Exception("Unsupported expo-protocol-version: $expoProtocolVersion")
      }
    }
  }

  @Throws(JSONException::class)
  fun getEmbeddedManifest(manifestJson: JSONObject, configuration: UpdatesConfiguration): BareUpdateManifest {
    return BareUpdateManifest.fromBareManifest(BareManifest(manifestJson), configuration)
  }
}
